Improved abstract camera

AhmetCanSolak opened the pull request:
Following issue #166 , this PR adds details that are agreed on to abstract camera class implementation. Tried to achieve a minimalist implementation in abstract class, as we see more opportunities with common functionality in different camera adapters we can revisit abstract class again.

YangLiujames reviewed (commented):
can we add a function get estimate of the readout+exposure time?

keithchev commented on copylot/hardware/cameras/abstract_camera.py:
```diff
@@ -1,5 +1,113 @@
+        """Method to capture a single image from the camera."""
+        raise NotImplementedError()
+
+    @abstractmethod
+    def list_available_cameras(self):
+        """List all cameras that driver discovers."""
+        raise NotImplementedError()
```
It feels surprising that this method would be an instance method of a particular camera (in other words, why is listing all cameras a function of `self`)?

AhmetCanSolak replied:
yes this could be static

keithchev commented on copylot/hardware/cameras/abstract_camera.py:
```diff
@@ -1,5 +1,113 @@
+
+    @image_size.setter
+    @abstractmethod
+    def image_size(self, value):
+        raise NotImplementedError()
```
should it be clarified here that `value` must be a tuple?

AhmetCanSolak replied:
hard to say at  this point, I rather wait and see APIs from more camera vendors on this then  decide on what type to enforce in abstract camera class

edyoshikun replied:
I think here we should pass a tuple and the function can handle those as needed depending on vendors. Most likely we would need 2dimensions anyways.

AhmetCanSolak replied:
okay then I will go with tuple for now

keithchev reviewed (commented):
General thoughts:
- would it make sense to include some method to read image data? 
- for clarity, it would be helpful to include type annotations for the abstract methods
- I think a docstring explaining what methods are or are not included in the base class would be helpful (e.g., if there is no method to read image data, the docstring could explain why)

AhmetCanSolak commented:
> General thoughts:
> 
> * would it make sense to include some method to read image data?

Reading image data from where to where exactly?

> * for clarity, it would be helpful to include type annotations for the abstract methods

We should add types where possible, but for many members, I think it is early to decide on types of parameters and return.

> * I think a docstring explaining what methods are or are not included in the base class would be helpful (e.g., if there is no method to read image data, the docstring could explain why)

General philosophy is pushing to abstract class as little as needed. Anything implemented by only a single vendor is not welcomed in the abstract class. Do you want to have a disclaimer about this in docstrings @keithchev ?

keithchev commented:
> > General thoughts:
> > 
> > * would it make sense to include some method to read image data?
> 
> Reading image data from where to where exactly?

Sorry this wasn't clear - I meant reading the data from the camera's image buffer, as a way to access the image acquired by a prior call to `snap`. This *feels* intuitively like a function that every camera would support in some form, but I could be wrong. 

> > * for clarity, it would be helpful to include type annotations for the abstract methods
> 
> We should add types where possible, but for many members, I think it is early to decide on types of parameters and return.

I see. that makes sense, but I think it would be good to mention this in the docstring with a `TODO: decide on types` or something. 

> 
> > * I think a docstring explaining what methods are or are not included in the base class would be helpful (e.g., if there is no method to read image data, the docstring could explain why)
> 
> General philosophy is pushing to abstract class as little as needed. Anything implemented by only a single vendor is not welcomed in the abstract class. Do you want to have a disclaimer about this in docstrings @keithchev ?

I mean, we are told that explicit is better than implicit, so 😉

edyoshikun commented on copylot/hardware/cameras/abstract_camera.py:
```diff
@@ -1,5 +1,114 @@
-class AbstractCamera(metaclass=abc.ABCMeta):
-    pass
+class AbstractCamera(metaclass=ABCMeta):
+
```
Do we need a `device_ID()`? Sometimes it's useful referencing the cameras based on the serial number, but hard to say if it should be part of the abstract camera class. I know serial number would be something that doesn't change regardless of connection type (i.e USB, COM port, etc). @AhmetCanSolak @YangLiujames

AhmetCanSolak replied:
Which camera providers expose serial number of cameras through their APIs?

edyoshikun replied:
basler, flir, teledyne photometrix, thorlabs. In my experience most cameras have that setting since it is easier to initialize a camera through the serial number than USB specially when you have multiple cameras connected.

AhmetCanSolak replied:
how about having a class member called `serial_number`? which  can be assigned in different ways in different child classes. What do you think @edyoshikun ?

AhmetCanSolak replied:
I added the device_id property @edyoshikun , will merge now but if you have any change request let me know. I can always do follow-up PRs.

AhmetCanSolak commented:
> > > General thoughts:
> > > 
> > > * would it make sense to include some method to read image data?
> > 
> > 
> > Reading image data from where to where exactly?
> 
> Sorry this wasn't clear - I meant reading the data from the camera's image buffer, as a way to access the image acquired by a prior call to `snap`. This _feels_ intuitively like a function that every camera would support in some form, but obvi I could be wrong (and/or too biased by Micromanager's API).
> 

I understand what you are saying, but I think cameras handle those buffers in quite different ways. I would suggest us to revisit adding such method later after we have several camera adapters in use and seeing how those handle this situation. For the meantime, maybe you can make an issue of it @keithchev ? For us to not forget about it

> > > * for clarity, it would be helpful to include type annotations for the abstract methods
> > 
> > 
> > We should add types where possible, but for many members, I think it is early to decide on types of parameters and return.
> 
> I see. that makes sense, but I think it would be good to mention this in the docstring with a `TODO: decide on types` or something.
> 

added some type annotations

> > > * I think a docstring explaining what methods are or are not included in the base class would be helpful (e.g., if there is no method to read image data, the docstring could explain why)
> > 
> > 
> > General philosophy is pushing to abstract class as little as needed. Anything implemented by only a single vendor is not welcomed in the abstract class. Do you want to have a disclaimer about this in docstrings @keithchev ?
> 
> I mean, we are told that explicit is better than implicit, so 😉

I added a class level  docstrings talks on this.
